src/script/sign.h

```diff
@@ -67,9 +67,283 @@ struct SignatureData {
     explicit SignatureData(const CScript& script) : scriptSig(script) {}
 };
 
+// Note: These constants are in reverse byte order because serialization uses LSB
+static const uint32_t PSBT_MAGIC_BYTES = 0x74627370;
+static const uint8_t PSBT_UNSIGNED_TX = 0x00;
+static const uint8_t PSBT_REDEEMSCRIPT = 0x01;
+static const uint8_t PSBT_WITNESSSCRIPT = 0x02;
+static const uint8_t PSBT_NON_WITNESS_UTXO = 0x03;
+static const uint8_t PSBT_WITNESS_UTXO = 0x04;
+static const uint8_t PSBT_PARTIAL_SIG = 0x05;
+static const uint8_t PSBT_SEPARATOR = 0x00; // A key that has length 0 is a separator and has no value. That is just 1 byte
+
+/** An structure for PSBTs which contain per input information */
+struct PartiallySignedInput
+{
+    CTransactionRef non_witness_utxo;
+    CTxOut witness_utxo;
+    std::map<CPubKey, std::vector<unsigned char>> partial_sigs;
+
+    PartiallySignedInput(): non_witness_utxo(), witness_utxo(), partial_sigs() {}
+    void SetNull();
+};
+
+/** A version of CTransaction with the PSBT format*/
+struct PartiallySignedTransaction
+{
+    CMutableTransaction tx;
+    std::map<uint160, CScript> redeem_scripts;
+    std::map<uint256, CScript> witness_scripts;
+    std::vector<PartiallySignedInput> inputs;
+
+    PartiallySignedTransaction();
+    PartiallySignedTransaction(CMutableTransaction& tx, std::map<uint160, CScript>& redeem_scripts, std::map<uint256, CScript>& witness_scripts, std::vector<PartiallySignedInput>& inputs);
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+
+        // magic bytes
+        s << PSBT_MAGIC_BYTES << (char)0xff; // psbt 0xff
+
+        // unsigned tx flag
+        WriteCompactSize(s, sizeof(PSBT_UNSIGNED_TX));
+        s << PSBT_UNSIGNED_TX;
+
+        // Write serialized tx to a stream
+        WriteCompactSize(s, ::GetSerializeSize(tx, SER_NETWORK, PROTOCOL_VERSION));
+        s << tx;
+        // Write redeem scripts and witness scripts
+        for (auto& entry : redeem_scripts) {
+            WriteCompactSize(s, entry.first.size() + sizeof(PSBT_REDEEMSCRIPT));
```

Comment:
This "write size, write header, write some other data" seems to be a recurring pattern. Perhaps there is a cleaner way to accomplish that (helper function that takes a header byte + some other type, and turns it into a vector)?